### PR TITLE
Add multigrid hierarchy functions

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Functions to set up a hierarchy of coarsened grids for the multigrid linear solver.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
@kidder and @fmahebert would you be willing to take this one?

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Hierarchy.hpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <unordered_set>
+#include <vector>
+
+#include "Domain/Structure/ElementId.hpp"
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/Hierarchy.hpp]: I think you only need a forward decl

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Hierarchy.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+#include "Domain/Structure/Side.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace LinearSolver::multigrid {
+
+template <size_t Dim>
+std::vector<std::array<size_t, Dim>> coarsen(
+    std::vector<std::array<size_t, Dim>> initial_refinement_levels) noexcept {
+  for (size_t block_id = 0; block_id < initial_refinement_levels.size();
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/Hierarchy.cpp]: could this just be a range-based for loop?

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Hierarchy.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+       ++block_id) {
+    for (size_t d = 0; d < Dim; ++d) {
+      auto& refinement_level =
+          gsl::at(initial_refinement_levels.at(block_id), d);
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/Hierarchy.cpp]: `levels.at` -> `levels[block_id]`

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Multigrid.hpp [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+/// Items related to the multigrid linear solver
```
Comment (nilsdeppe) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/Multigrid.hpp:7]: should this be in the `Linear solver` group?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Please squash directly :)

**Comment by nilsvu** [comment]
Rebased and squashed in those few changes. Cheers @nilsdeppe!

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Pull request merged**